dev/NumberBox/NumberBoxParser.cpp

```
@@ -0,0 +1,301 @@
+{
+    std::vector<MathToken> postfixTokens;
+    std::stack<MathToken> operatorStack;
+
+    bool error = false;
+
+    for (auto const token : infixTokens)
+    {
+        if (token.Type == MathTokenType::Numeric)
+        {
+            postfixTokens.push_back(token);
+        }
+        else if (token.Type == MathTokenType::Operator)
+        {
+            while (!operatorStack.empty())
+            {
+                const auto top = operatorStack.top();
+                if (top.Type != MathTokenType::Parenthesis && (GetPrecedenceValue(top.Char) >= GetPrecedenceValue(token.Char)))
+                {
+                    postfixTokens.push_back(operatorStack.top());
+                    operatorStack.pop();
+                }
+                else
+                {
+                    break;
+                }
+            }
+            operatorStack.push(token);
+        }
+        else if (token.Type == MathTokenType::Parenthesis)
+        {
+            if (token.Char == L'(')
+            {
+                operatorStack.push(token);
+            }
+            else
+            {
+                while (!operatorStack.empty() && operatorStack.top().Char != L'(')
+                {
+                    // Pop operators onto output until we reach a left paren
+                    postfixTokens.push_back(operatorStack.top());
+                    operatorStack.pop();
+                }
+
+                if (operatorStack.empty())
+                {
+                    // Broken parenthesis
+                    error = true;
+                    break;
+                }
+
+                // Pop left paren and discard
+                operatorStack.pop();
+            }
+        }
+    }
+
+    // Pop all remaining operators.
+    while (!operatorStack.empty())
+    {
+        if (operatorStack.top().Type == MathTokenType::Parenthesis)
+        {
+            // Broken parenthesis
+            error = true;
```

>error = true; [](start = 12, length = 13)

There is no way to correct for an error so instead of tracking state and continuing could we just early return the error empty vector?

---

I thought early returns were forbidden?

---

hmm, i think it is discouraged, but not returning here seems like it could be a waste, you might parse the remainder of a huge expression that you knew early on would be an error.

---
In reply to: [345448495](https://github.com/microsoft/microsoft-ui-xaml/pull/1602#discussion_r345448495) [](ancestors = 345448495)

---

I totally agree so if you think it's cool I'm going to do that.